### PR TITLE
Document Gradient Clipping during Manual Optimization

### LINKED ISSUE #14598: Document gradient clipping with manual optimization
## 📚 Documentation

On [this page](https://pytorch-lightning.readthedocs.io/en/stable/model/manual_optimization.html?highlight=manual%20optimiztion), we should probably document the clip_gradients method that you can use in manual optimization, since it's not handled through the Trainer arguments. 

The docstring for this method should also be improved to mention manual optimization:

https://github.com/Lightning-AI/lightning/blob/95374440ce9b4f499f83b0e6b8cee0bc715aa18b/src/pytorch_lightning/core/module.py#L1474-L1491

In conjunction, also clarify when to use `configure_gradient_clipping` and when not (manual vs automatic opt). 

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/Lightning-AI/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/Lightning-AI/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/Lightning-AI/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/Lightning-AI/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging PyTorch Lightning, Transformers, and Hydra.

cc @borda @rohitgr7 @Felonious-Spellfire @akihironitta @krshrimali

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?
Fixes #14598 

- [x] Updated docs with gradient clipping information
- [x] Updated docstrings of `clip_gradients()` function

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by shenoynikhil** [comment]
@awaelchli For some reason, I am not able to run `make docs` from this [link](https://github.com/Lightning-AI/lightning/tree/master/docs#building-docs). A reproducible example,
```bash
conda create -n pl python=3.9
conda activate pl
pip install lightning
git submodule update --init --recursive
make docs
```
This gives the following,
```bash
ERROR: Could not find a version that satisfies the requirement lai-sphinx-theme (from versions: none)
ERROR: No matching distribution found for lai-sphinx-theme
```

**Comment by awaelchli** [comment]
@shenoynikhil You would need this additional step:
```
pip install requirements/pytorch/docs.txt
```

And then you should be able to do:

```
cd docs/source-pytorch
make html
```

**Comment by shenoynikhil** [comment]
Thanks @awaelchli, that worked. Fixed the docs error. I guess it is ready to be reviewed.

File: docs/source-pytorch/model/manual_optimization.rst [hunk 1]
```diff
@@ -98,6 +98,36 @@ after every ``N`` steps, you can do as such.
...
+Gradient Clipping
+=================
+
+You can clip optimizer gradients during manual optimization similar to passing ``gradient_clip_val`` and
+``gradient_clip_algorithm`` argument in :ref:`Trainer <trainer>` during automatic optimization.
+To perform gradient clipping with one optimizer with manual optimization, you can do as such.
+
+.. testcode:: python
+
+    def __init__(self, gradient_clip_val=0.5, gradient_clip_algorithm="norm"):
```
Comment (Borda) [inline docs/source-pytorch/model/manual_optimization.rst:110]: can we add class for the syntax

Comment (shenoynikhil) [reply docs/source-pytorch/model/manual_optimization.rst]: Updated it.

File: docs/source-pytorch/model/manual_optimization.rst [hunk 2]
```diff
@@ -98,6 +98,39 @@ after every ``N`` steps, you can do as such.
             opt.step()
             opt.zero_grad()
 
+Gradient Clipping
+=================
+
+You can clip optimizer gradients during manual optimization similar to passing ``gradient_clip_val`` and
+``gradient_clip_algorithm`` argument in :ref:`Trainer <trainer>` during automatic optimization.
+To perform gradient clipping with one optimizer with manual optimization, you can do as such.
```
Comment (awaelchli) [inline docs/source-pytorch/model/manual_optimization.rst:106]:
```suggestion
You can clip optimizer gradients during manual optimization similar to passing the ``gradient_clip_val`` and
``gradient_clip_algorithm`` argument in :ref:`Trainer <trainer>` during automatic optimization.
To perform gradient clipping with one optimizer with manual optimization, you can do as such.
```

File: docs/source-pytorch/model/manual_optimization.rst [hunk 3]
```diff
@@ -98,6 +98,39 @@ after every ``N`` steps, you can do as such.
...
+            opt = self.optimizers()
+
+            # compute loss
+            loss = self.compute_loss(batch)
+
+            opt.zero_grad()
+            self.manual_backward(loss)
+            # clip gradients
+            self.clip_gradients(opt, self.gradient_clip_val, self.gradient_clip_algorithm)
+            opt.step()
```
Comment (awaelchli) [inline docs/source-pytorch/model/manual_optimization.rst:130]:
```suggestion
            self.manual_backward(loss)
            
            # clip gradients
            self.clip_gradients(opt, self.gradient_clip_val, self.gradient_clip_algorithm)
            
            opt.step()
```

I'm suggesting a bit of spacing to highlight the line we showcase

File: docs/source-pytorch/model/manual_optimization.rst [hunk 4]
```diff
@@ -98,6 +98,39 @@ after every ``N`` steps, you can do as such.
...
+
+.. warning::
+   * Note that overriding/implementing ``configure_gradient_clipping()`` in the case of Manual Optimization will not perform gradient clipping / will be ignored. Instead consider using ``self.gradient_clipping()`` manually like the example above.
```
Comment (awaelchli) [inline docs/source-pytorch/model/manual_optimization.rst:133]:
```suggestion
   * Note that ``configure_gradient_clipping()`` won't be called in Manual Optimization. Instead consider using ``self. clip_gradients()`` manually like in the example above.
```

File: src/pytorch_lightning/core/module.py [hunk 5]
```diff
@@ -1479,6 +1479,34 @@ def clip_gradients(
...
             gradient_clip_algorithm: The gradient clipping algorithm to use. Pass ``gradient_clip_algorithm="value"``
                 to clip by value, and ``gradient_clip_algorithm="norm"`` to clip by norm.
+
+        Note:
+            For manual optimization (``self.automatic_optimization = False``), if you want to use gradient clipping,
+            consider calling ``self.clip_gradients(opt, gradient_clip_val, gradient_clip_algorithm)`` as shown in the
+            example (with a single optimizer) below,
+
+        Example::
+            from pytorch_lightning import LightningModule
```
Comment (awaelchli) [inline src/pytorch_lightning/core/module.py:1489]: This example is great, but it goes into the user guide docs, which is what you have already done. So let's remove it from here, or shorten it to just the `training_step` method.

File: src/pytorch_lightning/core/module.py [hunk 6]
```diff
@@ -1479,6 +1479,34 @@ def clip_gradients(
             gradient_clip_val: The value at which to clip gradients.
             gradient_clip_algorithm: The gradient clipping algorithm to use. Pass ``gradient_clip_algorithm="value"``
                 to clip by value, and ``gradient_clip_algorithm="norm"`` to clip by norm.
+
+        Note:
+            For manual optimization (``self.automatic_optimization = False``), if you want to use gradient clipping,
+            consider calling ``self.clip_gradients(opt, gradient_clip_val, gradient_clip_algorithm)`` as shown in the
+            example (with a single optimizer) below,
```
Comment (awaelchli) [inline src/pytorch_lightning/core/module.py:1486]: I suggest to drop the note and instead adjust the descriptions above to keep this concise.

**Review by awaelchli — COMMENTED** [review]
Looks great already!

File: src/pytorch_lightning/core/module.py (hunk 6 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/core/module.py:1486]: What do you mean by adjust the description. Making it concise like this? Or after `Handles gradient clipping internally`?
```
"""Handles gradient clipping internally.

      Note:
            - Do not override this method. If you want to customize gradient clipping, consider using :meth:`configure_gradient_clipping` method.
            - For manual optimization (``self.automatic_optimization = False``), if you want to use gradient clipping, consider calling ``self.clip_gradients(opt, gradient_clip_val, gradient_clip_algorithm)`` manually in the training step.

        Args:
            optimizer: Current optimizer being used.
            gradient_clip_val: The value at which to clip gradients.
            gradient_clip_algorithm: The gradient clipping algorithm to use. Pass ``gradient_clip_algorithm="value"``
                to clip by value, and ``gradient_clip_algorithm="norm"`` to clip by norm.
        """
```

File: src/pytorch_lightning/core/module.py (hunk 5 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/core/module.py:1489]: Makes sense.

File: docs/source-pytorch/model/manual_optimization.rst [hunk 7]
```diff
@@ -98,6 +98,41 @@ after every ``N`` steps, you can do as such.
...
+            opt = self.optimizers()
+
+            # compute loss
+            loss = self.compute_loss(batch)
+
+            opt.zero_grad()
+            self.manual_backward(loss)
+
+            # clip gradients
+            self.clip_gradients(opt, self.gradient_clip_val, self.gradient_clip_algorithm)
```
Comment (carmocca) [inline docs/source-pytorch/model/manual_optimization.rst:130]: I would suggest simplifying the example by avoiding the instance attributes, which should make it a bit easier to read. Also including the keyword names for clarity

```suggestion
            self.clip_gradients(opt, gradient_clip_val=0.5, gradient_clip_algorithm="norm")
```

Same comment for the other files.

Comment (shenoynikhil) [reply docs/source-pytorch/model/manual_optimization.rst]: Updated it in both files.

File: src/pytorch_lightning/core/module.py (hunk 6 above)
Comment (shenoynikhil) [reply src/pytorch_lightning/core/module.py]: Done.

**Review by awaelchli — APPROVED** [review]
😃

File: src/pytorch_lightning/core/module.py [hunk 8]
```diff
@@ -1471,8 +1471,12 @@ def clip_gradients(
...
 
         Note:
-            Do not override this method. If you want to customize gradient clipping, consider
-            using :meth:`configure_gradient_clipping` method.
+            - Do not override this method. If you want to customize gradient clipping, consider using
+            :meth:`configure_gradient_clipping` method.
+            - For manual optimization (``self.automatic_optimization = False``), if you want to use
+            gradient clipping, consider calling
+            ``self.clip_gradients(opt, gradient_clip_val=0.5, gradient_clip_algorithm="norm")``
+            manually in the training step.
```
Comment (awaelchli) [inline src/pytorch_lightning/core/module.py:1479]: One more small issue: 

```
Just one more small issue to fix: /home/runner/work/lightning/lightning/src/pytorch_lightning/core/module.py:docstring of pytorch_lightning.core.module.LightningModule.clip_gradients:6: WARNING: Bullet list ends without a blank line; unexpected unindent.
```


```suggestion
            - Do not override this method. If you want to customize gradient clipping, consider using
              :meth:`configure_gradient_clipping` method.
            - For manual optimization (``self.automatic_optimization = False``), if you want to use
              gradient clipping, consider calling
              ``self.clip_gradients(opt, gradient_clip_val=0.5, gradient_clip_algorithm="norm")``
              manually in the training step.
```
